tensorflow/compiler/tf2tensorrt/convert/utils.h

```diff
@@ -128,77 +129,199 @@ inline bool IsTrtShapeTensorCompatible(const Tensor& tensor) {
+  DimsAdapter(int32_t num_dims, StorageType data)
+      : num_dims_(num_dims), storage_(std::forward<StorageType>(data)) {}
+
+  // Construct from a TRT dims object.
+  DimsAdapter(const nvinfer1::Dims& dims) : num_dims_(dims.nbDims) {
+    absl::c_copy(absl::MakeSpan(dims.d, dims.d + std::max(dims.nbDims, 0)),
+                 std::back_inserter(storage_));
+    LOG(INFO) << ::tensorflow::tensorrt::DebugString(dims) << "->"
+              << this->DebugString();
+  }
+
+  // Construct from a TensorShape or PartialTensorShape.
+  template <typename T>
+  explicit DimsAdapter(const TensorShapeBase<T>& shape,
+                       bool ignore_first_dim = false) {
+    LOG(INFO) << shape.DebugString();
+    if (shape.dims() > nvinfer1::Dims::MAX_DIMS)
+      status_ = errors::InvalidArgument("dims of TensorShape exceed MAX_DIMS");
+    if (ignore_first_dim && shape.dims() <= 0)
+      status_ = errors::InvalidArgument(
+          "removing first dim requires explicit batch dimension");
+    if (!status_.ok()) return;
+    if (shape.dims() == -1) {
+      *this = DimsAdapter(-1, StorageType{});
+      return;
     }
-    *trt_dims = {0, {1}};
-  } else {
-    const int offset = (ignore_first_dim ? 1 : 0);
-    for (int i = offset; i < shape.size(); i++) {
-      trt_dims->d[i - offset] = shape.at(i);
+    if (shape.dims() == 0) {
+      *this = DimsAdapter(0, StorageType{1});
+      return;
     }
-    trt_dims->nbDims = shape.size() - offset;
+    auto offt = (ignore_first_dim ? 1 : 0);
+    *this = DimsAdapter(
+        absl::MakeSpan(shape.dim_sizes().begin() + offt, shape.dims() - offt));
   }
-  return Status::OK();
-}
 
-template <typename TensorShapeType>
-nvinfer1::Dims TensorShapeToTrtDimsUnsafe(const TensorShapeType& tensor) {
-  nvinfer1::Dims dims;
-  dims.nbDims = tensor.dims();
-  for (int i = 0; i < dims.nbDims; i++) {
-    dims.d[i] = tensor.dim_size(i);
+  // Construct from an absl::Span.
+  template <typename T>
+  explicit DimsAdapter(absl::Span<T> shape)
+      : num_dims_(static_cast<int32_t>(shape.size())) {
+    absl::c_copy_n(shape, shape.size(), std::back_inserter(storage_));
   }
-  return dims;
-}
 
-template <typename TensorShapeType>
-Status TensorShapeToTrtDims(const TensorShapeType& shape, bool ignore_first_dim,
-                            nvinfer1::Dims* trt_dims) {
-  if (shape.dims() == -1) {
-    trt_dims->nbDims = -1;
+  // Construct from a container.
+  template <typename InputSequence,
+            typename = IsNotTensorShapeType<InputSequence>>
+  explicit DimsAdapter(const InputSequence& shape,
+                       bool ignore_first_dim = false)
+      : DimsAdapter(absl::MakeSpan(shape).subspan(ignore_first_dim ? 1 : 0,
+                                                  shape.size())) {}
+
+  // Construct from a status.
+  explicit DimsAdapter(Status status) : status_(std::move(status)) {}
+
+  //----- Conversion Utilities ------
+
+  // Convert to an nvinfer1::Dims and assign to the passed pointer.
```

Comment:
Let's revise this comment:
Converts to an nvinfers::Dims and assign the result to the object passed in via the result pointer.